controllers/user_controller.go

```diff
@@ -0,0 +1,252 @@
+				return fmt.Errorf("failed setting controller reference: %v", err)
+			}
+			return nil
+		})
+		return apiError
+	})
+	if err != nil {
+		msg := "failed to create/update credentials secret"
+		r.Recorder.Event(&credentialSecret, corev1.EventTypeWarning, string(operationResult), msg)
+		logger.Error(err, msg)
+		return err
+	}
+
+	logger.Info("Successfully declared credentials secret", "user", credentialSecret.ObjectMeta.Name)
+	r.Recorder.Event(&credentialSecret, corev1.EventTypeNormal, "SuccessfulDeclare", "Successfully declared user")
+	return nil
+}
+
+func (r *UserReconciler) setUserStatus(ctx context.Context, user *topologyv1alpha1.User) error {
+	logger := ctrl.LoggerFrom(ctx)
+
+	credentials := &corev1.LocalObjectReference{
+		Name: user.Spec.Name + "-user-credentials",
+	}
+	user.Status.Credentials = credentials
+	if err := r.Status().Update(ctx, user); err != nil {
+		return err
+	}
+	logger.Info("Successfully updated secret status credentials", "user", user.Name, "secretRef", credentials)
+	r.Recorder.Event(user, corev1.EventTypeNormal, "SuccessfulStatusUpdate", "Successfully updated user status")
+	return nil
+}
+
+func (r *UserReconciler) declareUser(ctx context.Context, client *rabbithole.Client, user *topologyv1alpha1.User) error {
+	logger := ctrl.LoggerFrom(ctx)
+
+	credentials := &corev1.Secret{}
+	if err := r.Get(ctx, types.NamespacedName{Name: user.Status.Credentials.Name, Namespace: user.Namespace}, credentials); err != nil {
+		return err
+	}
+
+	userSettings, err := internal.GenerateUserSettings(credentials, user.Spec.Tags)
+	if err != nil {
+		msg := "failed to generate user settings from credential in status"
+		r.Recorder.Event(user, corev1.EventTypeWarning, "FailedDeclare", msg)
+		logger.Error(err, msg, "user.status", user.Status)
+		return err
+	}
+
+	if err := validateResponse(client.PutUser(user.Spec.Name, userSettings)); err != nil {
+		msg := "failed to declare user"
+		r.Recorder.Event(user, corev1.EventTypeWarning, "FailedDeclare", msg)
+		logger.Error(err, msg, "user", user.Spec.Name)
+		return err
+	}
+
+	logger.Info("Successfully declared user", "user", user.Spec.Name)
+	r.Recorder.Event(user, corev1.EventTypeNormal, "SuccessfulDeclare", "Successfully declared user")
+	return nil
+}
+
+// addFinalizerIfNeeded adds a deletion finalizer if the User does not have one yet and is not marked for deletion
+func (r *UserReconciler) addFinalizerIfNeeded(ctx context.Context, user *topologyv1alpha1.User) error {
+	if user.ObjectMeta.DeletionTimestamp.IsZero() && !controllerutil.ContainsFinalizer(user, userFinalizer) {
+		controllerutil.AddFinalizer(user, userFinalizer)
+		if err := r.Client.Update(ctx, user); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *UserReconciler) deleteUser(ctx context.Context, client *rabbithole.Client, user *topologyv1alpha1.User) error {
+	logger := ctrl.LoggerFrom(ctx)
+
+	if err := validateResponse(client.DeleteUser(user.Spec.Name)); err != nil {
```

Comment:
You can use helper`validateResponseForDeletion()` https://github.com/rabbitmq/messaging-topology-operator/blob/main/controllers/utils.go#L117 here. We shouldn't requeue on already deleted users, same as queues and exchanges.